whylog/teacher/__init__.py

```diff
@@ -19,8 +20,7 @@ class Teacher(object):
 
     def __init__(self, id_to_line_dict, effect_id, config, pattern_assistant):
         self._lines = id_to_line_dict.copy()
-        self.effect_id = effect_id
-        self.constraints = set()
+        self.rule = UserRuleIntent(effect_id, parsers={}, constraints=[])
```

Comment:
`UserRuleIntent` should be instantiated just before being passed to Config. It is a DAO, which cannot comprehend things like history or relationships between items, or methods on them. The only reason this class exists is to pass it between Teacher and Config.